hs-ucan/library/Crypto/Key/Asymmetric/Public/Types.hs

```diff
@@ -1,70 +1,83 @@
+import qualified Crypto.PubKey.Ed25519                            as Crypto.Ed25519
+import qualified Crypto.PubKey.RSA                                as Crypto.RSA
+import qualified Crypto.Secp256k1
 
 import           Data.Swagger
 import           Database.Persist.Sql
 
-import           Control.Lens                                  ((?~))
-import           Data.Aeson
+import           Control.Lens                                     ((?~))
+import           Data.Aeson                                       as Json
+import qualified Data.Aeson.Types                                 as Json
 import           RIO
-import qualified RIO.Text                                      as Text
+import qualified RIO.ByteString.Lazy                              as Lazy
+import qualified RIO.Text                                         as Text
 import           Servant.API
 import           Test.QuickCheck
 
-import           Web.UCAN.Internal.RSA2048.Pair.Types          as Pair
+import           Web.UCAN.Internal.RSA2048.Pair.Types             as Pair
 
-import           Web.UCAN.Internal.Orphanage.Ed25519.PublicKey ()
-import           Web.UCAN.Internal.Orphanage.RSA2048.Public    ()
+import           Web.UCAN.Internal.Orphanage.Ed25519.PublicKey    ()
+import           Web.UCAN.Internal.Orphanage.RSA2048.Public       ()
+import           Web.UCAN.Internal.Orphanage.Secp256k1.PublicKey  ()
 
 data Public
-  = Ed25519PublicKey Crypto.Ed25519.PublicKey
-  | RSAPublicKey     Crypto.RSA.PublicKey
-  deriving Eq
+  = Ed25519PublicKey   Crypto.Ed25519.PublicKey
+  | RSAPublicKey       Crypto.RSA.PublicKey
+  | Secp256k1PublicKey Crypto.Secp256k1.PubKeyXY
+  deriving (Eq)
 
 instance Show Public where
   show = \case
-    Ed25519PublicKey ed -> Text.unpack $ textDisplay ed
-    RSAPublicKey     pk -> show pk
+    Ed25519PublicKey   ed -> Text.unpack $ textDisplay ed
+    RSAPublicKey       pk -> show pk
+    Secp256k1PublicKey pk -> Text.unpack $ textDisplay pk
 
 instance Display Public where
-  textDisplay (Ed25519PublicKey pk) = textDisplay pk
-  textDisplay (RSAPublicKey     pk) = textDisplay pk
+  textDisplay (Ed25519PublicKey   pk) = textDisplay pk
+  textDisplay (RSAPublicKey       pk) = textDisplay pk
+  textDisplay (Secp256k1PublicKey pk) = textDisplay pk
 
 instance Arbitrary Public where
   arbitrary = oneof
     [ Ed25519PublicKey <$> arbitrary
     , RSAPublicKey . Pair.pk <$> arbitrary
+    , Secp256k1PublicKey <$> arbitrary
     ]
 
 instance ToHttpApiData Public where
   toUrlPiece = textDisplay
 
 instance FromHttpApiData Public where
-  parseUrlPiece txt =
-    if "MII" `Text.isPrefixOf` txt
-      then RSAPublicKey     <$> parseUrlPiece txt
-      else Ed25519PublicKey <$> parseUrlPiece txt
+  parseUrlPiece text =
+    case (Json.decode . Lazy.fromStrict . Text.encodeUtf8 $ text) :: Maybe Public of
+      Just pub -> Right pub
+      Nothing -> Left "Failed to decode the JSON formatted public key"
```

Comment:
As far as I've seen we only use this in the fixtures.
Not sure if this has any other implications.

Reply:
Created new v3 routes to handle this.